query/src/plan/stringset.rs

```diff
@@ -0,0 +1,257 @@
+use std::sync::Arc;
+
+use arrow_deps::{datafusion::logical_plan::LogicalPlan, util::str_iter_to_batch};
+use data_types::TABLE_NAMES_COLUMN_NAME;
+
+use crate::{
+    exec::stringset::{StringSet, StringSetRef},
+    util::make_scan_plan,
+};
+
+use snafu::{ResultExt, Snafu};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Internal error converting to arrow: {}", source))]
+    InternalConvertingToArrow {
+        source: arrow_deps::arrow::error::ArrowError,
+    },
+
+    #[snafu(display("Internal creating a plan for stringset: {}", source))]
+    InternalPlanningStringSet {
+        source: arrow_deps::datafusion::error::DataFusionError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A plan which produces a logical set of Strings (e.g. tag
+/// values). This includes variants with pre-calculated results as
+/// well a variant that runs a full on DataFusion plan.
+#[derive(Debug)]
+pub enum StringSetPlan {
+    /// The plan errored and we are delaying reporting the results
+    KnownError(Box<dyn std::error::Error + Send + Sync + 'static>),
+
+    /// The results are known from metadta only without having to run
+    /// an actual datafusion plan
+    KnownOk(StringSetRef),
+
+    /// A DataFusion plan(s) to execute. Each plan must produce
+    /// RecordBatches with exactly one String column, though the
+    /// the values produced by the plan may be repeated
+    ///
+    /// TODO: it would be cool to have a single datafusion LogicalPlan
+    /// that merged all the results together. However, no such Union
+    /// node exists at the time of writing, so we do the unioning in IOx
+    Plan(Vec<LogicalPlan>),
+}
+
+impl From<StringSetRef> for StringSetPlan {
+    /// Create a StringSetPlan from a StringSetRef
+    fn from(set: StringSetRef) -> Self {
+        Self::KnownOk(set)
+    }
+}
+
+impl From<StringSet> for StringSetPlan {
+    /// Create a StringSetPlan from a StringSet result, wrapping the error type
+    /// appropriately
+    fn from(set: StringSet) -> Self {
+        Self::KnownOk(StringSetRef::new(set))
+    }
+}
+
+impl<E> From<Result<StringSetRef, E>> for StringSetPlan
+where
+    E: std::error::Error + Send + Sync + 'static,
+{
+    /// Create a StringSetPlan from a Result<StringSetRef> result, wrapping the
```

Comment:
I haven't commented everywhere you could be using code formatting in the comments, but this one seems redundant with the non-code text:

```suggestion
    /// Create a `StringSetPlan` from a `Result<StringSetRef>`, wrapping the
```